// ngx-image-cropper/bundles/ngx-image-cropper.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-image-cropper', ['exports', '@angular/core', '@angular/platform-browser', '@angular/common'], factory) :
    (factory((global['ngx-image-cropper'] = {}),global.ng.core,global.ng.platformBrowser,global.ng.common));
}(this, (function (exports,core,platformBrowser,common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /**
     * @param {?} srcBase64
     * @return {?}
     */
    function resetExifOrientation(srcBase64) {
        try {
            /** @type {?} */
            var exifRotation = getExifRotation(srcBase64);
            if (exifRotation > 1) {
                return transformBase64BasedOnExifRotation(srcBase64, exifRotation);
            }
            else {
                return Promise.resolve(srcBase64);
            }
        }
        catch (ex) {
            return Promise.reject(ex);
        }
    }
    /**
     * @param {?} srcBase64
     * @param {?} exifRotation
     * @return {?}
     */
    function transformBase64BasedOnExifRotation(srcBase64, exifRotation) {
        return new Promise(function (resolve, reject) {
            /** @type {?} */
            var img = new Image();
            img.onload = function () {
                /** @type {?} */
                var width = img.width;
                /** @type {?} */
                var height = img.height;
                /** @type {?} */
                var canvas = document.createElement('canvas');
                /** @type {?} */
                var ctx = canvas.getContext('2d');
                if (ctx) {
                    if (4 < exifRotation && exifRotation < 9) {
                        canvas.width = height;
                        canvas.height = width;
                    }
                    else {
                        canvas.width = width;
                        canvas.height = height;
                    }
                    transformCanvas(ctx, exifRotation, width, height);
                    ctx.drawImage(img, 0, 0);
                    resolve(canvas.toDataURL());
                }
                else {
                    reject(new Error('No context'));
                }
            };
            img.src = srcBase64;
        });
    }
    /**
     * @param {?} imageBase64
     * @return {?}
     */
    function getExifRotation(imageBase64) {
        /** @type {?} */
        var view = new DataView(base64ToArrayBuffer(imageBase64));
        if (view.getUint16(0, false) != 0xFFD8) {
            return -2;
        }
        /** @type {?} */
        var length = view.byteLength;
        /** @type {?} */
        var offset = 2;
        while (offset < length) {
            if (view.getUint16(offset + 2, false) <= 8)
                return -1;
            /** @type {?} */
            var marker = view.getUint16(offset, false);
            offset += 2;
            if (marker == 0xFFE1) {
                if (view.getUint32(offset += 2, false) != 0x45786966) {
                    return -1;
                }
                /** @type {?} */
                var little = view.getUint16(offset += 6, false) == 0x4949;
                offset += view.getUint32(offset + 4, little);
                /** @type {?} */
                var tags = view.getUint16(offset, little);
                offset += 2;
                for (var i = 0; i < tags; i++) {
                    if (view.getUint16(offset + (i * 12), little) == 0x0112) {
                        return view.getUint16(offset + (i * 12) + 8, little);
                    }
                }
            }
            else if ((marker & 0xFF00) != 0xFF00) {
                break;
            }
            else {
                offset += view.getUint16(offset, false);
            }
        }
        return -1;
    }
    /**
     * @param {?} imageBase64
     * @return {?}
     */
    function base64ToArrayBuffer(imageBase64) {
        imageBase64 = imageBase64.replace(/^data\:([^\;]+)\;base64,/gmi, '');
        /** @type {?} */
        var binaryString = atob(imageBase64);
        /** @type {?} */
        var len = binaryString.length;
        /** @type {?} */
        var bytes = new Uint8Array(len);
        for (var i = 0; i < len; i++) {
            bytes[i] = binaryString.charCodeAt(i);
        }
        return bytes.buffer;
    }
    /**
     * @param {?} ctx
     * @param {?} orientation
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    function transformCanvas(ctx, orientation, width, height) {
        switch (orientation) {
            case 2:
                ctx.transform(-1, 0, 0, 1, width, 0);
                break;
            case 3:
                ctx.transform(-1, 0, 0, -1, width, height);
                break;
            case 4:
                ctx.transform(1, 0, 0, -1, 0, height);
                break;
            case 5:
                ctx.transform(0, 1, 1, 0, 0, 0);
                break;
            case 6:
                ctx.transform(0, 1, -1, 0, height, 0);
                break;
            case 7:
                ctx.transform(0, -1, -1, 0, height, width);
                break;
            case 8:
                ctx.transform(0, -1, 1, 0, 0, width);
                break;
        }
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /*
     * Hermite resize - fast image resize/resample using Hermite filter.
     * https://github.com/viliusle/Hermite-resize
     */
    /**
     * @param {?} canvas
     * @param {?} width
     * @param {?} height
     * @param {?=} resizeCanvas
     * @return {?}
     */
    function resizeCanvas(canvas, width, height, resizeCanvas) {
        if (resizeCanvas === void 0) {
            resizeCanvas = true;
        }
        /** @type {?} */
        var width_source = canvas.width;
        /** @type {?} */
        var height_source = canvas.height;
        width = Math.round(width);
        height = Math.round(height);
        /** @type {?} */
        var ratio_w = width_source / width;
        /** @type {?} */
        var ratio_h = height_source / height;
        /** @type {?} */
        var ratio_w_half = Math.ceil(ratio_w / 2);
        /** @type {?} */
        var ratio_h_half = Math.ceil(ratio_h / 2);
        /** @type {?} */
        var ctx = canvas.getContext('2d');
        if (ctx) {
            /** @type {?} */
            var img = ctx.getImageData(0, 0, width_source, height_source);
            /** @type {?} */
            var img2 = ctx.createImageData(width, height);
            /** @type {?} */
            var data = img.data;
            /** @type {?} */
            var data2 = img2.data;
            for (var j = 0; j < height; j++) {
                for (var i = 0; i < width; i++) {
                    /** @type {?} */
                    var x2 = (i + j * width) * 4;
                    /** @type {?} */
                    var center_y = j * ratio_h;
                    /** @type {?} */
                    var weight = 0;
                    /** @type {?} */
                    var weights = 0;
                    /** @type {?} */
                    var weights_alpha = 0;
                    /** @type {?} */
                    var gx_r = 0;
                    /** @type {?} */
                    var gx_g = 0;
                    /** @type {?} */
                    var gx_b = 0;
                    /** @type {?} */
                    var gx_a = 0;
                    /** @type {?} */
                    var xx_start = Math.floor(i * ratio_w);
                    /** @type {?} */
                    var yy_start = Math.floor(j * ratio_h);
                    /** @type {?} */
                    var xx_stop = Math.ceil((i + 1) * ratio_w);
                    /** @type {?} */
                    var yy_stop = Math.ceil((j + 1) * ratio_h);
                    xx_stop = Math.min(xx_stop, width_source);
                    yy_stop = Math.min(yy_stop, height_source);
                    for (var yy = yy_start; yy < yy_stop; yy++) {
                        /** @type {?} */
                        var dy = Math.abs(center_y - yy) / ratio_h_half;
                        /** @type {?} */
                        var center_x = i * ratio_w;
                        /** @type {?} */
                        var w0 = dy * dy; //pre-calc part of w
                        for (var xx = xx_start; xx < xx_stop; xx++) {
                            /** @type {?} */
                            var dx = Math.abs(center_x - xx) / ratio_w_half;
                            /** @type {?} */
                            var w = Math.sqrt(w0 + dx * dx);
                            if (w >= 1) {
                                //pixel too far
                                continue;
                            }
                            //hermite filter
                            weight = 2 * w * w * w - 3 * w * w + 1;
                            /** @type {?} */
                            var pos_x = 4 * (xx + yy * width_source);
                            //alpha
                            gx_a += weight * data[pos_x + 3];
                            weights_alpha += weight;
                            //colors
                            if (data[pos_x + 3] < 255)
                                weight = weight * data[pos_x + 3] / 250;
                            gx_r += weight * data[pos_x];
                            gx_g += weight * data[pos_x + 1];
                            gx_b += weight * data[pos_x + 2];
                            weights += weight;
                        }
                    }
                    data2[x2] = gx_r / weights;
                    data2[x2 + 1] = gx_g / weights;
                    data2[x2 + 2] = gx_b / weights;
                    data2[x2 + 3] = gx_a / weights_alpha;
                }
            }
            //clear and resize canvas
            if (resizeCanvas) {
                canvas.width = width;
                canvas.height = height;
            }
            else {
                ctx.clearRect(0, 0, width_source, height_source);
            }
            //draw
            ctx.putImageData(img2, 0, 0);
        }
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var ImageCropperComponent = (function () {
        function ImageCropperComponent(sanitizer, cd, zone) {
            this.sanitizer = sanitizer;
            this.cd = cd;
            this.zone = zone;
            this.setImageMaxSizeRetries = 0;
            this.cropperScaledMinWidth = 20;
            this.cropperScaledMinHeight = 20;
            this.marginLeft = '0px';
            this.imageVisible = false;
            this.format = 'png';
            this.outputType = 'both';
            this.maintainAspectRatio = true;
            this.aspectRatio = 1;
            this.resizeToWidth = 0;
            this.cropperMinWidth = 0;
            this.cropperMinHeight = 0;
            this.roundCropper = false;
            this.onlyScaleDown = false;
            this.imageQuality = 92;
            this.autoCrop = true;
            this.cropper = {
                x1: -100,
                y1: -100,
                x2: 10000,
                y2: 10000
            };
            this.alignImage = 'center';
            this.startCropImage = new core.EventEmitter();
            this.imageCropped = new core.EventEmitter();
            this.imageCroppedBase64 = new core.EventEmitter();
            this.imageCroppedFile = new core.EventEmitter();
            this.imageLoaded = new core.EventEmitter();
            this.cropperReady = new core.EventEmitter();
            this.loadImageFailed = new core.EventEmitter();
            this.initCropper();
        }
        Object.defineProperty(ImageCropperComponent.prototype, "imageFileChanged", {
            set: /**
             * @param {?} file
             * @return {?}
             */ function (file) {
                this.initCropper();
                if (file) {
                    this.loadImageFile(file);
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ImageCropperComponent.prototype, "imageChangedEvent", {
            set: /**
             * @param {?} event
             * @return {?}
             */ function (event) {
                this.initCropper();
                if (event && event.target && event.target.files && event.target.files.length > 0) {
                    this.loadImageFile(event.target.files[0]);
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ImageCropperComponent.prototype, "imageBase64", {
            set: /**
             * @param {?} imageBase64
             * @return {?}
             */ function (imageBase64) {
                this.initCropper();
                this.checkExifAndLoadBase64Image(imageBase64);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} changes
         * @return {?}
         */
        ImageCropperComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                if (changes["cropper"]) {
                    this.setMaxSize();
                    this.setCropperScaledMinSize();
                    this.checkCropperPosition(false);
                    this.doAutoCrop();
                    this.cd.markForCheck();
                }
                if (changes["aspectRatio"] && this.imageVisible) {
                    this.resetCropperPosition();
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.initCropper = /**
         * @return {?}
         */
            function () {
                this.imageVisible = false;
                this.originalImage = null;
                this.safeImgDataUrl = 'data:image/png;base64,iVBORw0KGg'
                    + 'oAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAC0lEQVQYV2NgAAIAAAU'
                    + 'AAarVyFEAAAAASUVORK5CYII=';
                this.moveStart = {
                    active: false,
                    type: null,
                    position: null,
                    x1: 0,
                    y1: 0,
                    x2: 0,
                    y2: 0,
                    clientX: 0,
                    clientY: 0
                };
                this.maxSize = {
                    width: 0,
                    height: 0
                };
                this.originalSize = {
                    width: 0,
                    height: 0
                };
                this.cropper.x1 = -100;
                this.cropper.y1 = -100;
                this.cropper.x2 = 10000;
                this.cropper.y2 = 10000;
            };
        /**
         * @param {?} file
         * @return {?}
         */
        ImageCropperComponent.prototype.loadImageFile = /**
         * @param {?} file
         * @return {?}
         */
            function (file) {
                var _this = this;
                /** @type {?} */
                var fileReader = new FileReader();
                fileReader.onload = function (event) {
                    /** @type {?} */
                    var imageType = file.type;
                    if (_this.isValidImageType(imageType)) {
                        _this.checkExifAndLoadBase64Image(event.target.result);
                    }
                    else {
                        _this.loadImageFailed.emit();
                    }
                };
                fileReader.readAsDataURL(file);
            };
        /**
         * @param {?} type
         * @return {?}
         */
        ImageCropperComponent.prototype.isValidImageType = /**
         * @param {?} type
         * @return {?}
         */
            function (type) {
                return /image\/(png|jpg|jpeg|bmp|gif|tiff)/.test(type);
            };
        /**
         * @param {?} imageBase64
         * @return {?}
         */
        ImageCropperComponent.prototype.checkExifAndLoadBase64Image = /**
         * @param {?} imageBase64
         * @return {?}
         */
            function (imageBase64) {
                var _this = this;
                resetExifOrientation(imageBase64)
                    .then(function (resultBase64) { return _this.loadBase64Image(resultBase64); })
                    .catch(function () { return _this.loadImageFailed.emit(); });
            };
        /**
         * @param {?} imageBase64
         * @return {?}
         */
        ImageCropperComponent.prototype.loadBase64Image = /**
         * @param {?} imageBase64
         * @return {?}
         */
            function (imageBase64) {
                var _this = this;
                this.originalBase64 = imageBase64;
                this.safeImgDataUrl = this.sanitizer.bypassSecurityTrustResourceUrl(imageBase64);
                this.originalImage = new Image();
                this.originalImage.onload = function () {
                    _this.originalSize.width = _this.originalImage.width;
                    _this.originalSize.height = _this.originalImage.height;
                    _this.cd.markForCheck();
                };
                this.originalImage.src = imageBase64;
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.imageLoadedInView = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (this.originalImage != null) {
                    this.imageLoaded.emit();
                    this.setImageMaxSizeRetries = 0;
                    setTimeout(function () { return _this.checkImageMaxSizeRecursively(); });
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.checkImageMaxSizeRecursively = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (this.setImageMaxSizeRetries > 40) {
                    this.loadImageFailed.emit();
                }
                else if (this.sourceImage && this.sourceImage.nativeElement && this.sourceImage.nativeElement.offsetWidth > 0) {
                    this.setMaxSize();
                    this.setCropperScaledMinSize();
                    this.resetCropperPosition();
                    this.cropperReady.emit();
                    this.cd.markForCheck();
                }
                else {
                    this.setImageMaxSizeRetries++;
                    setTimeout(function () {
                        _this.checkImageMaxSizeRecursively();
                    }, 50);
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.onResize = /**
         * @return {?}
         */
            function () {
                this.resizeCropperPosition();
                this.setMaxSize();
                this.setCropperScaledMinSize();
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.rotateLeft = /**
         * @return {?}
         */
            function () {
                this.transformBase64(8);
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.rotateRight = /**
         * @return {?}
         */
            function () {
                this.transformBase64(6);
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.flipHorizontal = /**
         * @return {?}
         */
            function () {
                this.transformBase64(2);
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.flipVertical = /**
         * @return {?}
         */
            function () {
                this.transformBase64(4);
            };
        /**
         * @param {?} exifOrientation
         * @return {?}
         */
        ImageCropperComponent.prototype.transformBase64 = /**
         * @param {?} exifOrientation
         * @return {?}
         */
            function (exifOrientation) {
                var _this = this;
                if (this.originalBase64) {
                    transformBase64BasedOnExifRotation(this.originalBase64, exifOrientation)
                        .then(function (rotatedBase64) { return _this.loadBase64Image(rotatedBase64); });
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.resizeCropperPosition = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var sourceImageElement = this.sourceImage.nativeElement;
                if (this.maxSize.width !== sourceImageElement.offsetWidth || this.maxSize.height !== sourceImageElement.offsetHeight) {
                    this.cropper.x1 = this.cropper.x1 * sourceImageElement.offsetWidth / this.maxSize.width;
                    this.cropper.x2 = this.cropper.x2 * sourceImageElement.offsetWidth / this.maxSize.width;
                    this.cropper.y1 = this.cropper.y1 * sourceImageElement.offsetHeight / this.maxSize.height;
                    this.cropper.y2 = this.cropper.y2 * sourceImageElement.offsetHeight / this.maxSize.height;
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.resetCropperPosition = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var sourceImageElement = this.sourceImage.nativeElement;
                if (!this.maintainAspectRatio) {
                    this.cropper.x1 = 0;
                    this.cropper.x2 = sourceImageElement.offsetWidth;
                    this.cropper.y1 = 0;
                    this.cropper.y2 = sourceImageElement.offsetHeight;
                }
                else if (sourceImageElement.offsetWidth / this.aspectRatio < sourceImageElement.offsetHeight) {
                    this.cropper.x1 = 0;
                    this.cropper.x2 = sourceImageElement.offsetWidth;
                    /** @type {?} */
                    var cropperHeight = sourceImageElement.offsetWidth / this.aspectRatio;
                    this.cropper.y1 = (sourceImageElement.offsetHeight - cropperHeight) / 2;
                    this.cropper.y2 = this.cropper.y1 + cropperHeight;
                }
                else {
                    this.cropper.y1 = 0;
                    this.cropper.y2 = sourceImageElement.offsetHeight;
                    /** @type {?} */
                    var cropperWidth = sourceImageElement.offsetHeight * this.aspectRatio;
                    this.cropper.x1 = (sourceImageElement.offsetWidth - cropperWidth) / 2;
                    this.cropper.x2 = this.cropper.x1 + cropperWidth;
                }
                this.doAutoCrop();
                this.imageVisible = true;
            };
        /**
         * @param {?} event
         * @param {?} moveType
         * @param {?=} position
         * @return {?}
         */
        ImageCropperComponent.prototype.startMove = /**
         * @param {?} event
         * @param {?} moveType
         * @param {?=} position
         * @return {?}
         */
            function (event, moveType, position) {
                if (position === void 0) {
                    position = null;
                }
                event.preventDefault();
                this.moveStart = __assign({ active: true, type: moveType, position: position, clientX: this.getClientX(event), clientY: this.getClientY(event) }, this.cropper);
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.moveImg = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.moveStart.active) {
                    event.stopPropagation();
                    event.preventDefault();
                    if (this.moveStart.type === 'move') {
                        this.move(event);
                        this.checkCropperPosition(true);
                    }
                    else if (this.moveStart.type === 'resize') {
                        this.resize(event);
                        this.checkCropperPosition(false);
                    }
                    this.cd.detectChanges();
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.setMaxSize = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var sourceImageElement = this.sourceImage.nativeElement;
                this.maxSize.width = sourceImageElement.offsetWidth;
                this.maxSize.height = sourceImageElement.offsetHeight;
                this.marginLeft = this.sanitizer.bypassSecurityTrustStyle('calc(50% - ' + this.maxSize.width / 2 + 'px)');
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.setCropperScaledMinSize = /**
         * @return {?}
         */
            function () {
                if (this.originalImage) {
                    this.setCropperScaledMinWidth();
                    this.setCropperScaledMinHeight();
                }
                else {
                    this.cropperScaledMinWidth = 20;
                    this.cropperScaledMinHeight = 20;
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.setCropperScaledMinWidth = /**
         * @return {?}
         */
            function () {
                this.cropperScaledMinWidth = this.cropperMinWidth > 0
                    ? Math.max(20, this.cropperMinWidth / this.originalImage.width * this.maxSize.width)
                    : 20;
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.setCropperScaledMinHeight = /**
         * @return {?}
         */
            function () {
                if (this.maintainAspectRatio) {
                    this.cropperScaledMinHeight = Math.max(20, this.cropperScaledMinWidth / this.aspectRatio);
                }
                else if (this.cropperMinHeight > 0) {
                    this.cropperScaledMinHeight = Math.max(20, this.cropperMinHeight / this.originalImage.height * this.maxSize.height);
                }
                else {
                    this.cropperScaledMinHeight = 20;
                }
            };
        /**
         * @param {?=} maintainSize
         * @return {?}
         */
        ImageCropperComponent.prototype.checkCropperPosition = /**
         * @param {?=} maintainSize
         * @return {?}
         */
            function (maintainSize) {
                if (maintainSize === void 0) {
                    maintainSize = false;
                }
                if (this.cropper.x1 < 0) {
                    this.cropper.x2 -= maintainSize ? this.cropper.x1 : 0;
                    this.cropper.x1 = 0;
                }
                if (this.cropper.y1 < 0) {
                    this.cropper.y2 -= maintainSize ? this.cropper.y1 : 0;
                    this.cropper.y1 = 0;
                }
                if (this.cropper.x2 > this.maxSize.width) {
                    this.cropper.x1 -= maintainSize ? (this.cropper.x2 - this.maxSize.width) : 0;
                    this.cropper.x2 = this.maxSize.width;
                }
                if (this.cropper.y2 > this.maxSize.height) {
                    this.cropper.y1 -= maintainSize ? (this.cropper.y2 - this.maxSize.height) : 0;
                    this.cropper.y2 = this.maxSize.height;
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.moveStop = /**
         * @return {?}
         */
            function () {
                if (this.moveStart.active) {
                    this.moveStart.active = false;
                    this.doAutoCrop();
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.move = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                /** @type {?} */
                var diffX = this.getClientX(event) - this.moveStart.clientX;
                /** @type {?} */
                var diffY = this.getClientY(event) - this.moveStart.clientY;
                this.cropper.x1 = this.moveStart.x1 + diffX;
                this.cropper.y1 = this.moveStart.y1 + diffY;
                this.cropper.x2 = this.moveStart.x2 + diffX;
                this.cropper.y2 = this.moveStart.y2 + diffY;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.resize = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                /** @type {?} */
                var diffX = this.getClientX(event) - this.moveStart.clientX;
                /** @type {?} */
                var diffY = this.getClientY(event) - this.moveStart.clientY;
                switch (this.moveStart.position) {
                    case 'left':
                        this.cropper.x1 = Math.min(this.moveStart.x1 + diffX, this.cropper.x2 - this.cropperScaledMinWidth);
                        break;
                    case 'topleft':
                        this.cropper.x1 = Math.min(this.moveStart.x1 + diffX, this.cropper.x2 - this.cropperScaledMinWidth);
                        this.cropper.y1 = Math.min(this.moveStart.y1 + diffY, this.cropper.y2 - this.cropperScaledMinHeight);
                        break;
                    case 'top':
                        this.cropper.y1 = Math.min(this.moveStart.y1 + diffY, this.cropper.y2 - this.cropperScaledMinHeight);
                        break;
                    case 'topright':
                        this.cropper.x2 = Math.max(this.moveStart.x2 + diffX, this.cropper.x1 + this.cropperScaledMinWidth);
                        this.cropper.y1 = Math.min(this.moveStart.y1 + diffY, this.cropper.y2 - this.cropperScaledMinHeight);
                        break;
                    case 'right':
                        this.cropper.x2 = Math.max(this.moveStart.x2 + diffX, this.cropper.x1 + this.cropperScaledMinWidth);
                        break;
                    case 'bottomright':
                        this.cropper.x2 = Math.max(this.moveStart.x2 + diffX, this.cropper.x1 + this.cropperScaledMinWidth);
                        this.cropper.y2 = Math.max(this.moveStart.y2 + diffY, this.cropper.y1 + this.cropperScaledMinHeight);
                        break;
                    case 'bottom':
                        this.cropper.y2 = Math.max(this.moveStart.y2 + diffY, this.cropper.y1 + this.cropperScaledMinHeight);
                        break;
                    case 'bottomleft':
                        this.cropper.x1 = Math.min(this.moveStart.x1 + diffX, this.cropper.x2 - this.cropperScaledMinWidth);
                        this.cropper.y2 = Math.max(this.moveStart.y2 + diffY, this.cropper.y1 + this.cropperScaledMinHeight);
                        break;
                }
                if (this.maintainAspectRatio) {
                    this.checkAspectRatio();
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.checkAspectRatio = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var overflowX = 0;
                /** @type {?} */
                var overflowY = 0;
                switch (this.moveStart.position) {
                    case 'top':
                        this.cropper.x2 = this.cropper.x1 + (this.cropper.y2 - this.cropper.y1) * this.aspectRatio;
                        overflowX = Math.max(this.cropper.x2 - this.maxSize.width, 0);
                        overflowY = Math.max(0 - this.cropper.y1, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x2 -= (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y1 += (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                    case 'bottom':
                        this.cropper.x2 = this.cropper.x1 + (this.cropper.y2 - this.cropper.y1) * this.aspectRatio;
                        overflowX = Math.max(this.cropper.x2 - this.maxSize.width, 0);
                        overflowY = Math.max(this.cropper.y2 - this.maxSize.height, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x2 -= (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y2 -= (overflowY * this.aspectRatio) > overflowX ? overflowY : (overflowX / this.aspectRatio);
                        }
                        break;
                    case 'topleft':
                        this.cropper.y1 = this.cropper.y2 - (this.cropper.x2 - this.cropper.x1) / this.aspectRatio;
                        overflowX = Math.max(0 - this.cropper.x1, 0);
                        overflowY = Math.max(0 - this.cropper.y1, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x1 += (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y1 += (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                    case 'topright':
                        this.cropper.y1 = this.cropper.y2 - (this.cropper.x2 - this.cropper.x1) / this.aspectRatio;
                        overflowX = Math.max(this.cropper.x2 - this.maxSize.width, 0);
                        overflowY = Math.max(0 - this.cropper.y1, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x2 -= (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y1 += (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                    case 'right':
                    case 'bottomright':
                        this.cropper.y2 = this.cropper.y1 + (this.cropper.x2 - this.cropper.x1) / this.aspectRatio;
                        overflowX = Math.max(this.cropper.x2 - this.maxSize.width, 0);
                        overflowY = Math.max(this.cropper.y2 - this.maxSize.height, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x2 -= (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y2 -= (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                    case 'left':
                    case 'bottomleft':
                        this.cropper.y2 = this.cropper.y1 + (this.cropper.x2 - this.cropper.x1) / this.aspectRatio;
                        overflowX = Math.max(0 - this.cropper.x1, 0);
                        overflowY = Math.max(this.cropper.y2 - this.maxSize.height, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x1 += (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y2 -= (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.doAutoCrop = /**
         * @return {?}
         */
            function () {
                if (this.autoCrop) {
                    this.crop();
                }
            };
        /**
         * @param {?=} outputType
         * @return {?}
         */
        ImageCropperComponent.prototype.crop = /**
         * @param {?=} outputType
         * @return {?}
         */
            function (outputType) {
                if (outputType === void 0) {
                    outputType = this.outputType;
                }
                if (this.sourceImage.nativeElement && this.originalImage != null) {
                    this.startCropImage.emit();
                    /** @type {?} */
                    var imagePosition = this.getImagePosition();
                    /** @type {?} */
                    var width = imagePosition.x2 - imagePosition.x1;
                    /** @type {?} */
                    var height = imagePosition.y2 - imagePosition.y1;
                    /** @type {?} */
                    var cropCanvas = (document.createElement('canvas'));
                    cropCanvas.width = width;
                    cropCanvas.height = height;
                    /** @type {?} */
                    var ctx = cropCanvas.getContext('2d');
                    if (ctx) {
                        if (this.backgroundColor != null) {
                            ctx.fillStyle = this.backgroundColor;
                            ctx.fillRect(0, 0, width, height);
                        }
                        ctx.drawImage(this.originalImage, imagePosition.x1, imagePosition.y1, width, height, 0, 0, width, height);
                        /** @type {?} */
                        var output = { width: width, height: height, imagePosition: imagePosition, cropperPosition: __assign({}, this.cropper) };
                        /** @type {?} */
                        var resizeRatio = this.getResizeRatio(width);
                        if (resizeRatio !== 1) {
                            output.width = Math.round(width * resizeRatio);
                            output.height = this.maintainAspectRatio
                                ? Math.round(output.width / this.aspectRatio)
                                : Math.round(height * resizeRatio);
                            resizeCanvas(cropCanvas, output.width, output.height);
                        }
                        return this.cropToOutputType(outputType, cropCanvas, output);
                    }
                }
                return null;
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.getImagePosition = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var sourceImageElement = this.sourceImage.nativeElement;
                /** @type {?} */
                var ratio = this.originalSize.width / sourceImageElement.offsetWidth;
                return {
                    x1: Math.round(this.cropper.x1 * ratio),
                    y1: Math.round(this.cropper.y1 * ratio),
                    x2: Math.min(Math.round(this.cropper.x2 * ratio), this.originalSize.width),
                    y2: Math.min(Math.round(this.cropper.y2 * ratio), this.originalSize.height)
                };
            };
        /**
         * @param {?} outputType
         * @param {?} cropCanvas
         * @param {?} output
         * @return {?}
         */
        ImageCropperComponent.prototype.cropToOutputType = /**
         * @param {?} outputType
         * @param {?} cropCanvas
         * @param {?} output
         * @return {?}
         */
            function (outputType, cropCanvas, output) {
                var _this = this;
                switch (outputType) {
                    case 'file':
                        return this.cropToFile(cropCanvas)
                            .then(function (result) {
                            output.file = result;
                            _this.imageCropped.emit(output);
                            return output;
                        });
                    case 'both':
                        output.base64 = this.cropToBase64(cropCanvas);
                        return this.cropToFile(cropCanvas)
                            .then(function (result) {
                            output.file = result;
                            _this.imageCropped.emit(output);
                            return output;
                        });
                    default:
                        output.base64 = this.cropToBase64(cropCanvas);
                        this.imageCropped.emit(output);
                        return output;
                }
            };
        /**
         * @param {?} cropCanvas
         * @return {?}
         */
        ImageCropperComponent.prototype.cropToBase64 = /**
         * @param {?} cropCanvas
         * @return {?}
         */
            function (cropCanvas) {
                /** @type {?} */
                var imageBase64 = cropCanvas.toDataURL('image/' + this.format, this.getQuality());
                this.imageCroppedBase64.emit(imageBase64);
                return imageBase64;
            };
        /**
         * @param {?} cropCanvas
         * @return {?}
         */
        ImageCropperComponent.prototype.cropToFile = /**
         * @param {?} cropCanvas
         * @return {?}
         */
            function (cropCanvas) {
                var _this = this;
                return this.getCanvasBlob(cropCanvas)
                    .then(function (result) {
                    if (result) {
                        _this.imageCroppedFile.emit(result);
                    }
                    return result;
                });
            };
        /**
         * @param {?} cropCanvas
         * @return {?}
         */
        ImageCropperComponent.prototype.getCanvasBlob = /**
         * @param {?} cropCanvas
         * @return {?}
         */
            function (cropCanvas) {
                var _this = this;
                return new Promise(function (resolve) {
                    cropCanvas.toBlob(function (result) { return _this.zone.run(function () { return resolve(result); }); }, 'image/' + _this.format, _this.getQuality());
                });
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.getQuality = /**
         * @return {?}
         */
            function () {
                return Math.min(1, Math.max(0, this.imageQuality / 100));
            };
        /**
         * @param {?} width
         * @return {?}
         */
        ImageCropperComponent.prototype.getResizeRatio = /**
         * @param {?} width
         * @return {?}
         */
            function (width) {
                return this.resizeToWidth > 0 && (!this.onlyScaleDown || width > this.resizeToWidth)
                    ? this.resizeToWidth / width
                    : 1;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.getClientX = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                return event.clientX || event.touches && event.touches[0] && event.touches[0].clientX;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.getClientY = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                return event.clientY || event.touches && event.touches[0] && event.touches[0].clientY;
            };
        ImageCropperComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'image-cropper',
                        template: "<div>\n    <img\n        #sourceImage\n        class=\"source-image\"\n        [src]=\"safeImgDataUrl\"\n        [style.visibility]=\"imageVisible ? 'visible' : 'hidden'\"\n        (load)=\"imageLoadedInView()\"\n    />\n    <div class=\"cropper\"\n         *ngIf=\"imageVisible\"\n         [class.rounded]=\"roundCropper\"\n         [style.top.px]=\"cropper.y1\"\n         [style.left.px]=\"cropper.x1\"\n         [style.width.px]=\"cropper.x2 - cropper.x1\"\n         [style.height.px]=\"cropper.y2 - cropper.y1\"\n         [style.margin-left]=\"alignImage === 'center' ? marginLeft : null\"\n         [style.visibility]=\"imageVisible ? 'visible' : 'hidden'\"\n    >\n        <div\n                (mousedown)=\"startMove($event, 'move')\"\n                (touchstart)=\"startMove($event, 'move')\"\n                class=\"move\"\n        >&nbsp;</div>\n        <span\n                class=\"resize topleft\"\n                (mousedown)=\"startMove($event, 'resize', 'topleft')\"\n                (touchstart)=\"startMove($event, 'resize', 'topleft')\"\n        ><span class=\"square\"></span></span>\n        <span\n                class=\"resize top\"\n        ><span class=\"square\"></span></span>\n        <span\n                class=\"resize topright\"\n                (mousedown)=\"startMove($event, 'resize', 'topright')\"\n                (touchstart)=\"startMove($event, 'resize', 'topright')\"\n        ><span class=\"square\"></span></span>\n        <span\n                class=\"resize right\"\n        ><span class=\"square\"></span></span>\n        <span\n                class=\"resize bottomright\"\n                (mousedown)=\"startMove($event, 'resize', 'bottomright')\"\n                (touchstart)=\"startMove($event, 'resize', 'bottomright')\"\n        ><span class=\"square\"></span></span>\n        <span\n                class=\"resize bottom\"\n        ><span class=\"square\"></span></span>\n        <span\n                class=\"resize bottomleft\"\n                (mousedown)=\"startMove($event, 'resize', 'bottomleft')\"\n                (touchstart)=\"startMove($event, 'resize', 'bottomleft')\"\n        ><span class=\"square\"></span></span>\n        <span\n                class=\"resize left\"\n        ><span class=\"square\"></span></span>\n        <span\n                class=\"resize-bar top\"\n                (mousedown)=\"startMove($event, 'resize', 'top')\"\n                (touchstart)=\"startMove($event, 'resize', 'top')\"\n        ></span>\n        <span\n                class=\"resize-bar right\"\n                (mousedown)=\"startMove($event, 'resize', 'right')\"\n                (touchstart)=\"startMove($event, 'resize', 'right')\"\n        ></span>\n        <span\n                class=\"resize-bar bottom\"\n                (mousedown)=\"startMove($event, 'resize', 'bottom')\"\n                (touchstart)=\"startMove($event, 'resize', 'bottom')\"\n        ></span>\n        <span\n                class=\"resize-bar left\"\n                (mousedown)=\"startMove($event, 'resize', 'left')\"\n                (touchstart)=\"startMove($event, 'resize', 'left')\"\n        ></span>\n    </div>\n</div>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        styles: [":host{display:flex;position:relative;width:100%;max-width:100%;max-height:100%;overflow:hidden;padding:5px;text-align:center;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}:host>div{position:relative;width:100%}:host>div img.source-image{max-width:100%;max-height:100%}:host .cropper{position:absolute;display:flex;color:#53535c;background:0 0;touch-action:none;outline:rgba(255,255,255,.3) solid 100vw}:host .cropper:after{position:absolute;content:'';top:0;bottom:0;left:0;right:0;pointer-events:none;border:1px dashed;opacity:.75;color:inherit;z-index:1}:host .cropper .move{width:100%;cursor:move;border:1px solid rgba(255,255,255,.5)}:host .cropper .resize{position:absolute;display:inline-block;line-height:6px;padding:8px;opacity:.85;z-index:1}:host .cropper .resize .square{display:inline-block;background:#53535c;width:6px;height:6px;border:1px solid rgba(255,255,255,.5);box-sizing:content-box}:host .cropper .resize.topleft{top:-12px;left:-12px;cursor:nwse-resize}:host .cropper .resize.top{top:-12px;left:calc(50% - 12px);cursor:ns-resize}:host .cropper .resize.topright{top:-12px;right:-12px;cursor:nesw-resize}:host .cropper .resize.right{top:calc(50% - 12px);right:-12px;cursor:ew-resize}:host .cropper .resize.bottomright{bottom:-12px;right:-12px;cursor:nwse-resize}:host .cropper .resize.bottom{bottom:-12px;left:calc(50% - 12px);cursor:ns-resize}:host .cropper .resize.bottomleft{bottom:-12px;left:-12px;cursor:nesw-resize}:host .cropper .resize.left{top:calc(50% - 12px);left:-12px;cursor:ew-resize}:host .cropper .resize-bar{position:absolute;z-index:1}:host .cropper .resize-bar.top{top:-11px;left:11px;width:calc(100% - 22px);height:22px;cursor:ns-resize}:host .cropper .resize-bar.right{top:11px;right:-11px;height:calc(100% - 22px);width:22px;cursor:ew-resize}:host .cropper .resize-bar.bottom{bottom:-11px;left:11px;width:calc(100% - 22px);height:22px;cursor:ns-resize}:host .cropper .resize-bar.left{top:11px;left:-11px;height:calc(100% - 22px);width:22px;cursor:ew-resize}:host .cropper.rounded{outline-color:transparent}:host .cropper.rounded:after{border-radius:100%;box-shadow:0 0 0 100vw rgba(255,255,255,.3)}@media (orientation:portrait){:host .cropper{outline-width:100vh}:host .cropper.rounded:after{box-shadow:0 0 0 100vh rgba(255,255,255,.3)}}:host .cropper.rounded .move{border-radius:100%}"]
                    }] }
        ];
        /** @nocollapse */
        ImageCropperComponent.ctorParameters = function () {
            return [
                { type: platformBrowser.DomSanitizer },
                { type: core.ChangeDetectorRef },
                { type: core.NgZone }
            ];
        };
        ImageCropperComponent.propDecorators = {
            sourceImage: [{ type: core.ViewChild, args: ['sourceImage',] }],
            imageFileChanged: [{ type: core.Input }],
            imageChangedEvent: [{ type: core.Input }],
            imageBase64: [{ type: core.Input }],
            format: [{ type: core.Input }],
            outputType: [{ type: core.Input }],
            maintainAspectRatio: [{ type: core.Input }],
            aspectRatio: [{ type: core.Input }],
            resizeToWidth: [{ type: core.Input }],
            cropperMinWidth: [{ type: core.Input }],
            cropperMinHeight: [{ type: core.Input }],
            roundCropper: [{ type: core.Input }],
            onlyScaleDown: [{ type: core.Input }],
            imageQuality: [{ type: core.Input }],
            autoCrop: [{ type: core.Input }],
            backgroundColor: [{ type: core.Input }],
            cropper: [{ type: core.Input }],
            alignImage: [{ type: core.HostBinding, args: ['style.text-align',] }, { type: core.Input }],
            startCropImage: [{ type: core.Output }],
            imageCropped: [{ type: core.Output }],
            imageCroppedBase64: [{ type: core.Output }],
            imageCroppedFile: [{ type: core.Output }],
            imageLoaded: [{ type: core.Output }],
            cropperReady: [{ type: core.Output }],
            loadImageFailed: [{ type: core.Output }],
            onResize: [{ type: core.HostListener, args: ['window:resize',] }],
            moveImg: [{ type: core.HostListener, args: ['document:mousemove', ['$event'],] }, { type: core.HostListener, args: ['document:touchmove', ['$event'],] }],
            moveStop: [{ type: core.HostListener, args: ['document:mouseup',] }, { type: core.HostListener, args: ['document:touchend',] }]
        };
        return ImageCropperComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var ImageCropperModule = (function () {
        function ImageCropperModule() {
        }
        ImageCropperModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [
                            ImageCropperComponent
                        ],
                        exports: [
                            ImageCropperComponent
                        ]
                    },] }
        ];
        return ImageCropperModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.ImageCropperModule = ImageCropperModule;
    exports.ImageCropperComponent = ImageCropperComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
