// ngx-material-timepicker/esm2015/src/app/material-timepicker/directives/ngx-material-timepicker-theme.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from '@angular/core';
export class NgxMaterialTimepickerThemeDirective {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.element = elementRef.nativeElement;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.theme) {
            this.setTheme(this.theme);
        }
    }
    /**
     * @private
     * @param {?} theme
     * @return {?}
     */
    setTheme(theme) {
        for (const val in theme) {
            if (theme.hasOwnProperty(val)) {
                if (typeof theme[val] === 'string') {
                    for (const prop in theme) {
                        if (theme.hasOwnProperty(prop)) {
                            this.element.style.setProperty(`--${camelCaseToDash(prop)}`, theme[prop]);
                        }
                    }
                    return;
                }
                this.setTheme(theme[val]);
            }
        }
    }
}
NgxMaterialTimepickerThemeDirective.decorators = [
    { type: Directive, args: [{ selector: '[ngxMaterialTimepickerTheme]' },] }
];
/** @nocollapse */
NgxMaterialTimepickerThemeDirective.ctorParameters = () => [
    { type: ElementRef }
];
NgxMaterialTimepickerThemeDirective.propDecorators = {
    theme: [{ type: Input, args: ['ngxMaterialTimepickerTheme',] }]
};
if (false) {
    /** @type {?} */
    NgxMaterialTimepickerThemeDirective.prototype.theme;
    /**
     * @type {?}
     * @private
     */
    NgxMaterialTimepickerThemeDirective.prototype.element;
}
/**
 * @param {?} myStr
 * @return {?}
 */
function camelCaseToDash(myStr) {
    return myStr.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
}
